#3900: Add prod support for batch and channels

ruthreshx opened the pull request:
Added the batch and channel support for Prod op.

Working on H & W.

tt-aho commented on tt_eager/tt_dnn/op_library/prod/kernels/utils.hpp:
```diff
@@ -0,0 +1,298 @@
+
+#include "dataflow_api.h"
+
+void fill_cb_with_value(uint32_t cb_id, uint32_t value, int32_t num_of_elems = 1024) {
+    cb_reserve_back(cb_id, 1);
+    auto ptr = reinterpret_cast<uint16_t *>(get_write_ptr(cb_id));
+    for (int j = 0; j < num_of_elems; j++) {
+        ptr[j] = uint16_t(value >> 16);
+    }
+    cb_push_back(cb_id, 1);
+}
```
Any reason why we use int for num_of_elems/loop?

For pointers to l1 memory you should create them as
`volatile tt_l1_ptr std::uint16_t* ptr = (volatile tt_l1_ptr uint16_t*)(get_write_ptr(cb_id));`
This is to avoid a potential hang due to HW bug.

Can also store/use the u16 value like how you do it in later code
`const auto u16_value = uint16_t(value >> 16);`

VirdhatchaniKN replied:
Functions in this file are not called anywhere inside PROD. So I'll delete this file.

tt-aho commented on tt_eager/tt_dnn/op_library/prod/kernels/utils.hpp:
```diff
@@ -0,0 +1,298 @@
+
+    // mask_h
+    // first tile ptr
+    auto mask_h_ptr = reinterpret_cast<uint16_t *>(get_write_ptr(cb_mask_h_w));
```
Same comment on l1 ptr as above

VirdhatchaniKN replied:
Functions in this file are not called anywhere inside PROD. So I'll delete this file.

VirdhatchaniKN replied:
`generate_mask_h_w` function is not used.Hence, I'm removing this function

tt-aho commented on tt_eager/tt_dnn/op_library/prod/kernels/utils.hpp:
```diff
@@ -0,0 +1,298 @@
+            if (mask_h_0 >= 16) {
+                mask_h_0 = 16;
+            }
+            uint32_t h = 0;
+            for (; h < mask_h_0; h++) {
+                mask_h_ptr[h * 16 + w] = u16_one;
+            }
+            for (; h < 16; h++) {
+                mask_h_ptr[h * 16 + w] = u16_zero;
+            }
```
Avoid using multiplies

ruthreshx commented on tt_eager/tt_dnn/op_library/prod/prod_nc_op.cpp:
```diff
@@ -0,0 +1,120 @@
+
+operation::ProgramWithCallbacks Prod::create_program(
+    const std::vector<Tensor>& inputs, std::vector<Tensor>& outputs) const {
+    TT_ASSERT((dim >= 0 && dim <= 3), "dim should be 0 - 3");
```
Removed the assert statement.

ruthreshx commented on tt_eager/tt_dnn/op_library/prod/prod_nc_op.cpp:
```diff
@@ -0,0 +1,120 @@
+////////////////////////////////////////////////////////////////////////////
+void Prod::validate(const std::vector<Tensor>& inputs) const {
+    TT_ASSERT((dim >= 0 && dim <= 3), "dim should be 0 - 3");
+    const auto& input = inputs.at(0);
+    const auto& output = inputs.at(1);
```
Added the rank 4 assert in the code accordingly.

ruthreshx commented on tt_eager/tt_dnn/op_library/prod/prod_nc_op.cpp:
```diff
@@ -0,0 +1,120 @@
+
+    for (int i = 0; i < input_shape.rank(); ++i) {
+        TT_ASSERT(input_shape[i] == output_shape[i]);
+        TT_ASSERT(input_shape_wo_padding[i] == output_shape_wo_padding[i]);
```
Done

ruthreshx commented on tt_eager/tt_dnn/op_library/prod/prod_nc_op.hpp:
```diff
@@ -0,0 +1,53 @@
+    const int64_t &dim,
+    const MemoryConfig &mem_config = operation::DEFAULT_OUTPUT_MEMORY_CONFIG);
+
+Tensor prod(
+    const Tensor &input,
+    const Tensor &output,
+    std::vector<int64_t> &dims,
+    const MemoryConfig &mem_config = operation::DEFAULT_OUTPUT_MEMORY_CONFIG);
```
Mem_config requires for create_output_tensor function.

tt-aho replied:
So this is used for intermediate mem config then right? I think if you want to keep this arg you should change the arg name in the binding, since it is currently labelled as output_mem_config, but your output tensor already has a mem_config that is used for the final output.

ruthreshx commented:
Hi @tt-aho ,
Addressed the comments.
Reg the L1 Ptr and avoid using the multiples,
I took reference from moreh sum, to added the support for N & C to implement the prod.
I just replicate the same thing for the prod as well.

TT-BrianLiu commented on tt_eager/tt_dnn/op_library/prod/kernels/prod_hw.cpp:
```diff
@@ -0,0 +1,62 @@
+#include "compute_kernel_api/reduce.h"
+#include "compute_kernel_api/tile_move_copy.h"
+
+ALWI void ACQ() { acquire_dst(tt::DstMode::Half); }
+ALWI void REL() { release_dst(tt::DstMode::Half); }
```
Switch to new APIs:
- tile_regs_commit, tile_regs_release
- tile_regs_acquire, tile_regs_wait

Docs: tt_metal/include/compute_kernel_api/reg_api.h

TT-BrianLiu commented on tests/tt_eager/python_api_testing/unit_testing/test_prod.py:
```diff
@@ -0,0 +1,93 @@
+        "8, 8, TILE_HEIGHT * 4 - 1, TILE_WIDTH * 4 - 1",
+    ],
+)
+@pytest.mark.parametrize(
+    "dims",
+    (
+        [
+            0,
+        ],
+        [
+            1,
+        ],
+        [
+            2,
+        ],
+        [
+            3,
+        ],
+    ),
+    ids=["0", "1", "2", "3"],
+)
```
fix this

TT-BrianLiu commented on tt_eager/tt_dnn/op_library/prod/kernels/reader_prod_hw.cpp:
```diff
@@ -0,0 +1,114 @@
+
+#include "dataflow_api.h"
+
+void mask_tile_in_reader(uint32_t l1_addr, uint32_t mask_w = 32, uint32_t mask_h = 32) {
+    union {
+        float f;
+        uint32_t u;
+    } zero;
+    zero.f = 0.0f;
+    auto ptr = reinterpret_cast<uint16_t *>(l1_addr);
+    for (uint32_t h = 0; h < 16; h++) {
```
do we have to do this in the kernel?

TT-BrianLiu commented on tt_eager/tt_dnn/op_library/prod/kernels/reader_prod_hw.cpp:
```diff
@@ -0,0 +1,114 @@
+    constexpr uint32_t cb_id_in1 = 1;
+    constexpr uint32_t cb_id_in2 = 2;
+    cb_reserve_back(cb_id_in2, 1);
+    if (scaler != 0) {
+        auto ptr = reinterpret_cast<uint16_t *>(get_write_ptr(cb_id_in2));
+        for (int j = 0; j < 1024; j++) ptr[j] = uint16_t(0);
+
+        for (int k = 0; k < 4; k++)
+            for (int j = 0; j < 16; j++) ptr[k * 256 + j] = uint16_t(scaler >> 16);
+    }
```
Do we have to do this in kernel?

TT-BrianLiu commented:
> Filling CBs is done in the kernel by others including MOREH in recent code. I don't see anything objectionable @brian Liu ***@***.***>
> […](#)
> On Fri, Feb 9, 2024 at 9:31 AM TT-BrianLiu ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In tt_eager/tt_dnn/op_library/prod/kernels/prod_hw.cpp <[#4933 (comment)](https://github.com/tenstorrent-metal/tt-metal/pull/4933#discussion_r1484608292)> : > +ALWI void ACQ() { acquire_dst(tt::DstMode::Half); } +ALWI void REL() { release_dst(tt::DstMode::Half); } Switch to new APIs: - tile_regs_commit, tile_regs_release - tile_regs_acquire, tile_regs_wait Docs: tt_metal/include/compute_kernel_api/reg_api.h ------------------------------ In tests/tt_eager/python_api_testing/unit_testing/test_prod.py <[#4933 (comment)](https://github.com/tenstorrent-metal/tt-metal/pull/4933#discussion_r1484608569)> : > ***@***.***( + "dims", + ( + [ + 0, + ], + [ + 1, + ], + [ + 2, + ], + [ + 3, + ], + ), + ids=["0", "1", "2", "3"], +) fix this ------------------------------ In tt_eager/tt_dnn/op_library/prod/kernels/reader_prod_hw.cpp <[#4933 (comment)](https://github.com/tenstorrent-metal/tt-metal/pull/4933#discussion_r1484609628)> : > +void mask_tile_in_reader(uint32_t l1_addr, uint32_t mask_w = 32, uint32_t mask_h = 32) { + union { + float f; + uint32_t u; + } zero; + zero.f = 0.0f; + auto ptr = reinterpret_cast<uint16_t *>(l1_addr); + for (uint32_t h = 0; h < 16; h++) { do we have to do this in the kernel? ------------------------------ In tt_eager/tt_dnn/op_library/prod/kernels/reader_prod_hw.cpp <[#4933 (comment)](https://github.com/tenstorrent-metal/tt-metal/pull/4933#discussion_r1484611045)> : > + if (scaler != 0) { + auto ptr = reinterpret_cast<uint16_t *>(get_write_ptr(cb_id_in2)); + for (int j = 0; j < 1024; j++) ptr[j] = uint16_t(0); + + for (int k = 0; k < 4; k++) + for (int j = 0; j < 16; j++) ptr[k * 256 + j] = uint16_t(scaler >> 16); + } Do we have to do this in kernel? — Reply to this email directly, view it on GitHub <[#4933 (review)](https://github.com/tenstorrent-metal/tt-metal/pull/4933#pullrequestreview-1872878117)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/BAGOCNFAYHJAL4UNJHOHISTYSZMOFAVCNFSM6AAAAABCKYHGGGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZSHA3TQMJRG4> . You are receiving this because you commented.Message ID: ***@***.***>

We should look into better ways of lowering constants like this. Convs do this with small sharded tensors. If you want to leave it like this, please document it better and clean it up (for example what are those hardcoded numbers and why are loop indices ints)

VirdhatchaniKN commented on tt_eager/tt_dnn/op_library/prod/kernels/utils.hpp:
```diff
@@ -0,0 +1,298 @@
+    cb_push_back(cb_mask_h_w, 2);
+}
+
+void generate_mask_w(uint32_t cb_mask, uint32_t mask_w) {
```
`generate_mask_w ` function is not used.Hence, I'm removing this function
